sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java

```diff
@@ -0,0 +1,691 @@
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        context.output(csvMapper.mapRow(context.element()));
+      }
+    }
+
+    public static class CleanTmpFilesFromGcsFn extends DoFn<Object, Object> {
+      private final String bucketName;
+      private final String bucketPath;
+      private final SnowflakeCloudProvider snowFlakeCloudProvider;
+
+      public CleanTmpFilesFromGcsFn(
+          String bucketName, String bucketPath, SnowflakeCloudProvider snowFlakeCloudProvider) {
+        this.bucketName = bucketName;
+        this.bucketPath = bucketPath;
+        this.snowFlakeCloudProvider = snowFlakeCloudProvider;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) {
+        snowFlakeCloudProvider.removeFiles(bucketName, bucketPath);
+      }
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      if (getQuery() != null) {
+        builder.add(DisplayData.item("query", getQuery()));
+      }
+      if (getTable() != null) {
+        builder.add(DisplayData.item("table", getTable()));
+      }
+      builder.add(DisplayData.item("integrationName", getIntegrationName()));
+      builder.add(DisplayData.item("stagingBucketName", getStagingBucketName()));
+      builder.add(DisplayData.item("csvMapper", getCsvMapper().getClass().getName()));
+      builder.add(DisplayData.item("coder", getCoder().getClass().getName()));
+      if (getDataSourceProviderFn() instanceof HasDisplayData) {
+        ((HasDisplayData) getDataSourceProviderFn()).populateDisplayData(builder);
+      }
+    }
+  }
+
+  /**
+   * A POJO describing a {@link DataSource}, providing all properties allowing to create a {@link
+   * DataSource}.
+   */
+  @AutoValue
+  public abstract static class DataSourceConfiguration implements Serializable {
+    @Nullable
+    public abstract ValueProvider<String> getUrl();
+
+    @Nullable
+    public abstract ValueProvider<String> getUsername();
+
+    @Nullable
+    public abstract ValueProvider<String> getPassword();
+
+    @Nullable
+    public abstract ValueProvider<PrivateKey> getPrivateKey();
+
+    @Nullable
+    public abstract ValueProvider<String> getOauthToken();
+
+    @Nullable
+    public abstract ValueProvider<String> getDatabase();
+
+    @Nullable
+    public abstract ValueProvider<String> getWarehouse();
+
+    @Nullable
+    public abstract ValueProvider<String> getSchema();
+
+    @Nullable
+    public abstract ValueProvider<String> getServerName();
+
+    @Nullable
+    public abstract ValueProvider<Integer> getPortNumber();
+
+    @Nullable
+    public abstract ValueProvider<String> getRole();
+
+    @Nullable
+    public abstract ValueProvider<String> getAuthenticator();
+
+    @Nullable
+    public abstract ValueProvider<Integer> getLoginTimeout();
+
+    @Nullable
+    public abstract ValueProvider<Boolean> getSsl();
```

Comment:
In Read and Write methods we stopped using `ValueProvider ` maybe here we also don't need it ?

Reply:
💯